### ISSUE TITLE
oslogin breaks OpenSSL setup

### ISSUE THREAD

**Issue by azrle** [GoogleCloudPlatform/compute-image-packages#562]
Hi,

The problem has been shown that when we use `getpwent()` with oslogin nss module, SSL setup will be broken after the call.

Here is a simple code to reproduce it.
```c
#include <stdio.h>
#include <openssl/evp.h>
#include <pwd.h>

int main() {
    OpenSSL_add_all_algorithms();
    EVP_get_digestbyname("sha256") ? printf("loaded\n") : printf("not loaded\n");

    struct passwd *pw;
    setpwent();
    while ((pw = getpwent()) != 0) {}

    EVP_get_digestbyname("sha256") ? printf("loaded\n") : printf("not loaded\n");
    return 0;
}
```
```
$ gcc test-getpwent.c -lssl -lcrypto -o test-getpwent
$ ./test-getpwent
loaded
not loaded

# after removing oslogin from /etc/nsswitch.conf
$ ./test-getpwent
loaded
loaded
```

After tracking it down, I found that the cause can be the `curl_global_cleanup()` called by `oslogin_utils::HttpGet`.

As `oslogin` communicates with metadata without SSL/TLS, can we clear the `CURL_GLOBAL_SSL` bit when initialize curl so that it will not cleanup OpenSSL things?

A simple test is as follows.
```c
#include <stdio.h>
#include <openssl/evp.h>
#include <curl/curl.h>

int main() {
    OpenSSL_add_all_algorithms();
    EVP_get_digestbyname("sha256") ? printf("loaded\n") : printf("not loaded\n");

    curl_global_init(CURL_GLOBAL_ALL & ~CURL_GLOBAL_SSL);
    curl_global_cleanup();

    EVP_get_digestbyname("sha256") ? printf("loaded\n") : printf("not loaded\n");
    return 0;
}
```
```
$ gcc test-curl.c -lssl -lcrypto -lcurl -o test-curl
$ ./test-curl
loaded
loaded
```

Test environment:
```
libcurl3/xenial-updates,xenial-security,now 7.47.0-1ubuntu2.6 amd64 [installed]
libcurl3-gnutls/xenial-updates,xenial-security,now 7.47.0-1ubuntu2.6 amd64 [installed]
libcurl4-openssl-dev/xenial-updates,xenial-security,now 7.47.0-1ubuntu2.6 amd64 [installed]
openssl/xenial-updates,xenial-security,now 1.0.2g-1ubuntu4.10 amd64 [installed]
```

What do you think?

By the way, the original problem was shown when setting up instances via Chef on GCE with Ubuntu 16.04.  Ruby initializes the OpenSSL when requiring `openssl` (`Init_openssl`) and Chef calls `Etc.passwd` somewhere which leads to a runtime error when trying to use openssl again. 
```
FATAL: RuntimeError: cookbook_file[xxxx] (xxx::xxx line N) had an error: RuntimeError: Unsupported digest algorithm (SHA256).
```

**Comment by illfelder** [comment]
/cc @Sarsate 

Thanks for the report! We'll investigate and get back to you.

**Comment by wonderfly** [comment]
Bug was reported with curl 4.47, but in later versions (since 7.57) the constant `CURL_GLOBAL_SSL` has been deprecated and SSL is initialized and cleaned up unconditionally.

https://github.com/curl/curl/blob/master/include/curl/curl.h#L2543
https://github.com/curl/curl/blob/master/lib/easy.c#L331

We might need a different fix.

**Comment by zmarano** [comment]
Seems others are having this discussion as well:
https://github.com/curl/curl/issues/2276

**Comment by illfelder** [comment]
The current plan of action - we will move forward with adding the constant. This should unbreak SSL issues  on distros using older curl packages. The constant is treated as a no-op on newer versions, so we don't have to worry about the deprecation causing a new breakage. We will also work on replacing our usage of curl by using sockets directly. When implemented, this will fix the issue on newer images and prevent other similar problems from arising.

**Comment by wonderfly** [comment]
Doing http completely on your own sounds like a lot of work, but if you only have to implement a very small subset of it then it might be fine.

An alternative is to try and push libcurl upstream to fix the regression, especially since others have the same request.

**Comment by azrle** [comment]
Thanks for your quick fix!

**Comment by flokli** [comment]
This can be closed, #563 was merged.
